### ISSUE TITLE
[v6] setSearchParams called from parent erases child param from url

### ISSUE THREAD

**Issue by cdock1029** [ReactTraining/react-router#7704]
## Version

`6.0.0-beta.0`

## Test Case

https://codesandbox.io/s/react-router-forked-8tle3?file=/index.js

see: `toggleI` and `parentToggleI` in the `Topic` component

## Steps to reproduce

Expected behavior occurs when calling `setSearchParams` from within a child component matched by the current path param.

If `setSearchParams` is called from a parent, the existing matched child param is erased from the URL & we are navigated away from the currently rendered child route

## Expected Behavior

I'd expect only the search params to be modified. Navigation should maintain the current path and rendered descendant routes.

## Actual Behavior

The path limited to the parent route replaces whatever descendant routes were matched before `setSearchParams` was called.

**Comment by timdorr** [comment]
You should never pass `setSearchParams` as a prop. The hook is always intended to be directly invoked and anything returned by it is intended to be relative to the currently rendering component.

**Comment by cdock1029** [comment]
> You should never pass `setSearchParams` as a prop

Yes, sorry about that, this was a hastily constructed example used simply to demonstrate the real issue..

> anything returned by it is intended to be relative to the currently rendering component

Why? I would argue the hierarchy of route components / path segments in the url is orthogonal to query parameters, which can impact routes all along the hierarchy, and should be able to be updated at any point in the hierarchy without wiping away descendant routes. If they were *not* orthogonal, this would imply query params must necessarily be coupled to routes. Is this the intention?

The specific use case I'm dealing with is a data table, where expanded rows in in the table are child routes of the parent table route. Query parameters control transformations of the table, like sorting and grouping of columns. The transformations should occur independently from routing between different rows. The parent route holds the overall table transformation controls (clicking on a column header to sort, filtering, grouping etc).

Here's a simplified example:

https://codesandbox.io/s/react-router-forked-6ztp1?file=/index.js 

Basically, the behavior I'm thinking of would [modify this line](https://github.com/ReactTraining/react-router/blob/262b45dd8d6cbdc3a984259c5591ffc3de80d600/packages/react-router-dom/index.tsx#L418) to something like:

```ts
navigate(location.pathname + '?' + createSearchParams(nextInit), navigateOptions);
```
because I'd expect `setSearchParams` to just set the search params.

@mjackson
